pdcupdater/handlers/modules.py

```diff
@@ -0,0 +1,267 @@
+import os
+import logging
+import errno
+import re
+from subprocess import check_call, STDOUT
+
+import beanbag
+import pdcupdater.handlers
+import pdcupdater.services
+from pdcupdater.utils import TmpDir, PushPopD
+
+import modulemd
+
+log = logging.getLogger(__name__)
+
+
+class ModuleStateChangeHandler(pdcupdater.handlers.BaseHandler):
+    """ When the state of a module changes. """
+
+    relevant_states = ('done', 'ready')
+    valid_states = relevant_states + ('init', 'wait', 'failed')
+
+    tree_id_re = re.compile(
+        r"(?P<name>[^-]+)-(?P<version>[^-]+)-"
+        r"(?P<date>[0-9]{8})\.(?P<spin>[0-9]+)")
+
+    rpm_fname_re = re.compile(
+        r"(?P<name>.+)-"
+        r"(?:(?P<epoch>[^:]+):)?(?P<version>[^-:]+)-(?P<release>[^-]+)."
+        r"(?P<arch>[^\.]+).rpm")
+
+    scmurl_re = re.compile(
+        r"(?P<giturl>(?:(?P<scheme>git)://(?P<host>[^/]+))?"
+        r"(?P<repopath>/[^\?]+))\?(?P<modpath>[^#]*)#(?P<revision>.+)")
+
+    @property
+    def topic_suffixes(self):
+        return ['module.state.change']
+
+    def can_handle(self, msg):
+        if not msg['topic'].endswith('module.state.change'):
+            return False
+
+        state = msg['msg']['state']
+
+        if state not in self.valid_states:
+            log.warn("Invalid module state '{}', skipping.".format(state))
```

Comment:
Maybe make this a ``log.error`` statement instead.

The reason is that (in the production environment) warnings are sent to the journal, but errors are sent to the journal and are also emailed to developers/sysadmins.  If we get a message with an invalid state, something is *very* wrong and we want to know about it.